enhancements/E-0001-secret-handling.md

```diff
@@ -0,0 +1,748 @@
+nitty-gritty.
+-->
+
+- A new object of kind `SecretSync` will be introduced to the in-cluster API of PKO
+  - allowing a specified `Secret` source object to be synced to multiple specified namespace/name combinations
+  - implementing status reporting so that it can be used as a building block in automations - for example:
+    - PKO phases waiting for successful synchronization
+    - (Optional - feasibility unverified) PKO probes reflecting degraded workload health when a synced secret vanishes or during out-of-sync times after a source secret has been changed and the resynchronization hasn't happened yet
+  - usage can either happen
+    - freely (without using PKO's other API functionalities), maybe solving other problems of users not taken into further consideration in this proposal other than thinking about the performance implications on PKO and the apiserver this might have
+    - within packages with an additional guardrail (dubbed `secret-sync-fence`):
+      - By default, PKO checks that a `SecretSync` object that is templated by a package in the templating phase references one of the secrets specified in the `(Cluster)Package` object controlling the `(Cluster)ObjectDeployment` about to be created or updated to avoid accidentally syncing the wrong secrets
+      - This mechanism should have an opt-out hatch if one has good reasons to include specific `SecretSync` objects with hardcoded sources.
+      - This mechanism will probably turn out to be validatable at packaging- AND run-time. Validating at packaging-time would left-shift surfacing of guardrail violations and save a lot of iteration time!
+      - Important: this guardrail cannot be used for defensive security against malicious packagers because it would be easy to put an evil SecretSync object into a wrapper object like an OpenShift `Template` (or another `Package`), as PKO cannot possibly understand and correctly detect and unpack all possible forms of a "trojan horse" object.
+        - personal recommendation from Josh: If you want to be sure about the authenticity of a package image, implement validation of sigstore signatures on package images.
+
+- The PackageManifest API has to be changed so that `Secret` references can be specified and aliased. References specify:
+  - an alias name to be used with newly introduced resolving functions in the package templates
+  - when package is installable cluster-scoped: a namespaced name of the source secret
+  - when package is installable namespace-scoped: a name of the source secret which is expected in the Package's target namespace
+  - ~if the reference is optional or not (`TODO`: clarify implications - probably better off to start without optional secrets and introduce them later when needed. Big point: how to handle optionality without polling for secret's to appear? The package reconciler doesn't declaratively re-template the ObjectDeployment except, when .spec.image changes)~
```

Comment:
I think we never want - on this level - to watch for inputs to be created or changed.
Optional should probably mean that this references is not mandatory to be supplied.

If a secret is referenced, I think that the secret must exist to successfully unpack the image.
If the secret is removed after the package has been unpacked, we _could_ raise a new Condition on the Package to point it out to the user and make debugging easier.

Reply:
> I think we never want - on this level - to watch for inputs to be created or changed.
> Optional should probably mean that this references is not mandatory to be supplied.

Got it.
Would you argue _for_ including optional references in this proposal?

> If a secret is referenced, I think that the secret must exist to successfully unpack the image.

This is not as easy. The general pattern in kube is "if a secret reference isn't there then the affected controllers will keep retrying until it is there", to not take away forward- and backward-compat by requiring a specific thing to exist before it is used.

I explicitly propose _not_ checking for the existence of referenced secrets at unpack time and let the actual workloads requiring those secrets fail later on, which will be reflected by workload status and pko's probes. This way we minimize stepping on the toes of deployers and their deployment flows.